core/include/vecmem/containers/impl/static_array.ipp

```diff
@@ -0,0 +1,237 @@
+/**
+ * VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+#include "vecmem/utils/types.hpp"
+
+#include <cstddef>
+#include <type_traits>
+#include <stdexcept>
+
+namespace vecmem {
+    template<typename T, std::size_t N>
+    VECMEM_HOST_AND_DEVICE
+    static_array<T, N>::static_array(
+        void
+    ) {
+        /*
+         * This function does quite literally nothing, and leaves the array's
+         * contents uninitialized.
+         */
+    }
+
+    template<typename T, std::size_t N>
+    VECMEM_HOST
+    constexpr
+    typename static_array<T, N>::reference
+    static_array<T, N>::at(
+        size_type i
+    ) {
+        /*
+         * The at function is bounds-checking in the standard library, so we
+         * do a boundary check in our code, too. This makes this method
+         * incompatible with device code.
+         */
+        if (i >= N) {
+            throw std::out_of_range("Index greater than size of static array.");
+        }
```

Comment:
I'm not a fan of this. In all the other classes I forewent using exceptions, and instead just used assertions. Exactly so that they could be used in device code.

Let's keep it like this for now, but I believe we'll want to change this later on. As those assertions, I believe, will come in handy in device code once we start debugging more complicated algorithms. (While the advantage of exceptions will be pretty marginal in host code during those debugging sessions.)